src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1076 @@
+    // - new_lo[x] = ones[x], for x > i    (2.7)
+    // This is the result we are looking for.
+    //           1 1 0 1 0 0 1 0
+    new_lo |= bits._ones;
+    // Note that in this case, new_lo is always a valid answer. That is, it is
+    // a value not less than lo and satisfies bits.
+    assert(lo < new_lo, "the result must be valid");
+    return new_lo;
+  } else {
+    assert(zero_violation > one_violation, "remaining case");
+    // This means that the first bit that does not satisfy the bit requirement
+    // is a 1 that should be a 0.
+    //
+    // From section 3 above, we know i is the largest bit index such that:
+    // - lo[x] satisfies bits for 0 <= x < i (3.1)
+    // - zeros[i] = 0                        (3.2)
+    // - lo[i] = 0                           (3.3)
+    //
+    // We know that lo satisfies all bits before first_violation, hence (3.1)
+    // holds. However, first_violation is not the value i we are looking for
+    // because lo[first_violation] == 1. We can also see that any larger value
+    // of i would violate (3.1) since lo[first_violation] does not satisfy
+    // bits. As a result, we should find the last index x upto first_violation
+    // such that lo[x] == zeros[x] == 0. That value of x would be the value of
+    // i we are looking for.
+    //
+    // E.g:      1 2 3 4 5 6 7 8
+    //      lo = 1 0 0 0 1 1 1 0
+    //   zeros = 0 0 0 1 0 1 0 0
+    //    ones = 1 0 0 0 0 0 1 1
+    //   1-vio = 0 0 0 0 0 0 0 1
+    //   0-vio = 0 0 0 0 0 1 0 0
+    // The first violation is the 6th bit, which should be 0. We want to flip
+    // it to 0. However, since we must obtain a value larger than lo, we must
+    // find an earlier bit that can be flipped from 0 to 1. The 5th cannot be
+    // the bit we are looking for, because it is already 1, the 4th bit also
+    // cannot be, because it must be 0. As a result, the last bit we can flip,
+    // which is the first different bit between the result and lo must be the
+    // 3rd bit. As a result, the result must not be smaller than:
+    //           1 0 1 0 0 0 0 0
+    // This one satisfies zeros so we can use the logic in the previous case,
+    // just OR with ones to obtain the final result, which is:
+    //           1 0 1 0 0 0 1 1
+
+    juint first_violation = count_leading_zeros(zero_violation);
+    // This masks out all bits after the first violation
+    //           1 1 1 1 1 0 0 0
+    U find_mask = ~(std::numeric_limits<U>::max() >> first_violation);
+    // We want to find the last index x upto first_violation such that
+    // lo[x] == zeros[x] == 0.
+    // We start with all bits where lo[x] == zeros[x] == 0:
+    //           0 1 1 0 0 0 0 1
+    U either = lo | bits._zeros;
+    // Now let us find all the bit indices x upto first_violation such that
+    // lo[x] == zeros[x] == 0. The last one of these bits must be at index i.
+    //           0 1 1 0 0 0 0 0
+    U tmp = ~either & find_mask;
```

Comment:
To me a variable name `tmp` smells a little. I prefer expressive names. Up to you :)

Reply:
I find a name `neither_upto_first_violation` which seems more expressive, I'm not sure if it may raise any confusion.